### ISSUE TITLE
[EKS] [Feature]: Allow Kube Scheduler Customization

### ISSUE THREAD

**Issue by Kausheel** [aws/containers-roadmap#1468]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Tell us about your request**
What do you want us to build?

It would be great if EKS allowed users to configure the Kube Scheduler parameters. This is a Control Plane component, so users don't have access to this by default. Exposing the Kube Scheduler configuration either via AWS APIs or via the KubeSchedulerConfiguration resource type would be a significant advantage for EKS users.

**Which service(s) is this request for?**
EKS

**Tell us about the problem you're trying to solve. What are you trying to do, and why is it hard?**

Use cases for this might include switching from equal Pod distribution to a binpacking approach, which optimizes cost effectiveness. There are many other Scheduler parameters which users might want to tweak themselves.

**Are you currently working around this issue?**
Implementing custom Kube Schedulers. This is not ideal, since it requires operational overhead in maintaining and updating the custom Kube Scheduler. It may also require using tools like OPA to insert custom schedulerName fields into the target Pods, which is yet another burden on the user.

Thanks!

**Comment by ashishapy** [comment]
Another use-case is to define Cluster-level default constraints for PodTopologySpread in scheduler. As per doc https://kubernetes.io/docs/concepts/workloads/pods/pod-topology-spread-constraints/#cluster-level-default-constraints

AWS should make it as default behaviour in EKS cluster. 

`apiVersion: kubescheduler.config.k8s.io/v1beta1
kind: KubeSchedulerConfiguration
profiles:
  - pluginConfig:
      - name: PodTopologySpread
        args:
          defaultConstraints:
            - maxSkew: 1
              topologyKey: topology.kubernetes.io/zone
              whenUnsatisfiable: ScheduleAnyway
          defaultingType: List`

**Comment by stijndehaes** [comment]
I would love to use this for enabling bin packing like explained here:
https://kubernetes.io/docs/concepts/scheduling-eviction/resource-bin-packing/

**Comment by sherifabdlnaby** [comment]
Upvote. 

Trying to use EKS and achieve bin packing is hard without changing Scheduler Behavior to favor `MostAllocated`.

**Comment by logyball** [comment]
Note that this feature is [supported to some extent in Azure](https://github.com/Azure/aks-engine/blob/master/docs/topics/clusterdefinitions.md#schedulerconfig) and is supported for the use case of [Scheduler Scoring Strategy: MostAllocated](https://kubernetes.io/docs/reference/scheduling/config/) in [GKE by using the autoscaling profile](https://cloud.google.com/kubernetes-engine/docs/concepts/cluster-autoscaler#:~:text=The%20optimize%2Dutilization%20autoscaling%20profile,custom%20scheduler%20are%20not%20affected) (note this is an assumption on my part, GKE does not explicitly document _what_ this setting does under the hood) .  Adding this ability would help EKS users gain parity in that sense.

**Comment by stijndehaes** [comment]
I would be fine with having a setting like GKE has, this would solve my use case. It probably does not solve every use case out there, but I can understand if the AWS EKS team feels reluctant to allow changing the whole configuration.

**Comment by anhqqt** [comment]
Same here. We need this feature to enable resource bin packing for cost saving
https://kubernetes.io/docs/concepts/scheduling-eviction/resource-bin-packing/

**Comment by Art3mK** [comment]
@AnhQKatalon, run scheduler yourself with needed settings + patch pods to use that scheduler with kyverno for example :) Could be done in couple hours.

**Comment by anhqqt** [comment]
> @AnhQKatalon, run scheduler yourself with needed settings + patch pods to use that scheduler with kyverno for example :) Could be done in couple hours.

Yeah, I am doing the workaround this way. Appreciate your help. But it should be great if EKS supports changing the schedule configuration officially.

**Comment by AnhQKatalon** [comment]
> @AnhQKatalon, run scheduler yourself with needed settings + patch pods to use that scheduler with kyverno for example :) Could be done in couple hours.

Yeah, I am doing the workaround this way. Appreciate your help. But it should be great if EKS supports changing the schedule configuration officially.

**Comment by babinos87** [comment]
As others mentioned, this is required to set default pod topology constraints on the cluster, as per: https://kubernetes.io/docs/concepts/scheduling-eviction/topology-spread-constraints/#cluster-level-default-constraints. There would be other uses cases, I am sure of it.

There are workarounds, of course, but this seems like a core thing to do, in order to make the life of EKS users easier. I thing this is a MUST.

**Comment by fernandesnikhil** [comment]
This would be very helpful for the same reasons mentioned by other above:
- set default topology spreads for all pods in one central place
- tweak bin-packing by changing `NodeResourcesFit`

The suggestion of rolling your own Scheduler is not appealing because EKS might have bolted on their own tweaks/modifications to get the scheduler to work right in AWS and then we'd loose all of that. I get that modifying the EKS blessed set of configuration can lead to instability - but if I want to modify _just a few_ settings I should be allowed to do that with the understanding it could break scheduling on my cluster. Upstream k8s allows it and it's useful.

**Comment by subhranil05** [comment]
If not possible to add customization in kube-scheduler, 
can we think about this feature like GKE, node groups will have option to scale with  the `mostAllocated` like strategy like GKE have autoscale profile `optimize-utilization` ?

**Comment by sherifabdlnaby** [comment]
@subhranil05 This is not an alternative solution. Scaling Node Groups can only achieve bin-packing during the event of scaling up.  Kube Scheduler customization is necessary for in-place, proactive bin-packing.

**Comment by m00lecule** [comment]
Can somebody take a look and include this ticket to kanban board?

**Comment by paulchambers** [comment]
This would be very useful for my EKS clusters. I want to be able to set sensible defaults without having to run my own scheduler.

**Comment by cskinfill** [comment]
I would love to see this as well too support bin packing at scheduling.

**Comment by sherifabdlnaby** [comment]
Do it for the environment folks!

**Comment by Legion2** [comment]
I want to use bin packing with karpenter for job workloads. So karpenter can scale down empty nodes after a scale up. Instead of spreading the pods across all nearly empty nodes they should be packed on some full nodes, to enable karpenter removing empty nodes after the last job running on it completed.

**Comment by hfuss** [comment]
Assuming AWS may not prioritize this for awhile at the current rate, I think an example deployment of a custom scheduler with `MostAllocated` enabled for binpacking would benefit everyone here (as suggested in https://github.com/aws/containers-roadmap/issues/1468#issuecomment-1645021158) - despite the burden it puts on 1) cluster admins to maintain control plane infra in-step with EKS versions, 2) Pod creators to ensure the custom scheduler is used.

https://kubernetes.io/docs/tasks/extend-kubernetes/configure-multiple-schedulers/

Is a starting point, but if anyone has manifest samples for the exact configuration everyone wants that'd be appreciated. If I get to this at some point will share.

**Comment by vinay92-ch** [comment]
We ran into this this same issue and had to setup a [custom scheduler to implement bin-packing](https://clickhouse.com/blog/packing-kubernetes-pods-more-efficiently-saving-money?utm_medium=social&utm_source=twitter&utm_campaign=blog). It's the same kube-scheduler image with a custom scoring policy as suggested above. Blog more details about how we dealt with overprovisioning and system workloads and rollout to all pods. Relevant section from the blog with [scheduler](https://clickhouse.com/blog/packing-kubernetes-pods-more-efficiently-saving-money?utm_medium=social&utm_source=twitter&utm_campaign=blog#custom-scheduler-with-most-allocated-scoring-policy) config. 

We were able to achieve this in GCP by using the `optimize-utilization` setting in GKE, but for Azure AKS, we still have to use this secondary scheduler with custom scoring policy.

**Comment by MattLJoslin** [comment]
How is this API not supported yet? Is there any plan to support this soon? It's part of the standard Kubernetes service but there's no way to use on EKS? This really doesn't make EKS very usable in our case. All of the major packages are assuming that the standard APIs are available.

**Comment by eliran-zada-zesty** [comment]
Same as @MattLJoslin said... we really need it as well

**Comment by stevehipwell** [comment]
I think being able to run the scheduler in `MostAllocated` mode would make the Karpenter use case even more compelling.

**Comment by stevehipwell** [comment]
https://www.cncf.io/blog/2024/06/03/tackling-gpu-underutilization-in-kubernetes-runtimes/

**Comment by jukie** [comment]
Any updates on this?

**Comment by nikimanoledaki** [comment]
+1 for being able to add a `pluginConfig` about `PodTopologySpread` as well as a `MostAllocated` scoring policy.

> Do it for the environment folks!

And this!

**Comment by woehrl01** [comment]
Hint in the meantime, you can use the following AWS managed image to provision the scheduler yourself, without the need to self-manage the image: https://gallery.ecr.aws/eks-distro/kubernetes/kube-scheduler

**Comment by woehrl01** [comment]
@jukie I'm not arguing against that this would be a nice addition. I just wanted to mention a solution which won't require you to wait more than 3 years. Just wanted to share this to newcomers which maybe are having troubles with self maintaining this image, etc.

**Comment by jukie** [comment]
Totally agree, and thanks for sharing!

**Comment by gavinclarkeuk** [comment]
I raised this with our AWS support contacts, particularly highlighting the environmental impact of this issue and it looks like they might be taking it a bit more seriously. Suggest everyone does the same to see if we can build a bit more momentum behind this.

**Comment by HarshMehra** [comment]
Upvote,
Achieving bin packing in EKS is difficult without adjusting the scheduler's behavior to prioritize MostAllocated, which can contribute to cost savings on cloud servers. However, deploying a self-managed scheduler would introduce additional maintenance overhead, making the process more complex.

**Comment by zip-chanko** [comment]
Just check is this feature already available here https://github.com/aws-samples/custom-scheduler-eks?

**Comment by jukie** [comment]
No, that example is running a second scheduler inside the cluster which has some drawbacks and adds complexity overhead.

**Comment by theadzik** [comment]
Hello from 2026. We still need this feature.

https://kubernetes.io/docs/concepts/scheduling-eviction/resource-bin-packing/

**Comment by jpastoor** [comment]
Seems to have become a first-class EKS feature now 🥳 https://aws.amazon.com/about-aws/whats-new/2026/08/amazon-eks-control-plane-configuration-parameters/ with `kubeSchedulerConfig.nodeResourcesFit.scoringStrategy.type: MostAllocated`

**Comment by gavinclarkeuk** [comment]
Whilst I'd have preferred being able to provide a full scheduler ConfigMap, this is definitely a big improvement

**Comment by georgejohnis** [comment]
Closing this issue since users can now customize Kubernetes scheduler configuration including pod distribution (eg: binpacking) using EKS APIs. Learn more:
-   [What's new post](https://aws.amazon.com/about-aws/whats-new/2026/08/amazon-eks-control-plane-configuration-parameters/ ) 
-  [EKS User guide](https://docs.aws.amazon.com/eks/latest/userguide/control-plane-configuration.html)
